compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -363,20 +366,55 @@ fn loose_check<'cx, 'tcx>(
     selcx: &mut SelectionContext<'cx, 'tcx>,
     o: &PredicateObligation<'tcx>,
 ) -> bool {
+    // FIXME: the call to `selcx.predicate_may_hold_fatal` below should be ported
+    // to the canonical trait query form, `infcx.predicate_may_hold`, once
+    // the new system supports intercrate mode (which coherence needs).
     !selcx.predicate_may_hold_fatal(o)
 }
 
 fn negative_impl_exists<'cx, 'tcx>(
     selcx: &SelectionContext<'cx, 'tcx>,
+    region_context: DefId,
     o: &PredicateObligation<'tcx>,
 ) -> bool {
-    let infcx = selcx.infcx();
+    let infcx = selcx.infcx().clone();
```

Comment:
instead of deriving clone, let's make a method called `fork` and leave a comment like:

"Forks the inference context, creating a new inference context with the same inference variables in the same state. This can be used to "branch off" many tests from the same common state. Used in coherence."

This is a bit more explicit about intent than just calling `clone`, and will help to ensure that when new fields are added, the person adding them things a bit about what the right thing is to do. =)